lib/wasix/src/state/linker.rs

```diff
@@ -0,0 +1,751 @@
+                indirect_function_table: indirect_function_table.clone(),
+            })),
+        };
+
+        Ok((
+            linker,
+            LinkedMainModule {
+                memory,
+                indirect_function_table,
+                stack_low,
+                stack_high,
+            },
+        ))
+    }
+
+    // Initialize the linker with the instantiated main module. This needs to happen before the
+    // linker can be used to load any side modules.
+    pub fn initialize(&self, main_instance: Instance) {
+        let mut guard = self.state.lock().unwrap();
+        guard.main_instance = Some(main_instance);
+    }
+
+    // TODO: figure out how this should work with threads...
+    // TODO: auto-load dependencies, store relationship so dlsym can look inside deps of this lib as well
+    // TODO: give loaded library a different wasi env that specifies its module handle
+    // TODO: add ref-counting for already-loaded modules, so dlclose can know when to actually close a module
+    // TODO: call destructors
+    pub async fn load_module(
+        &self,
+        module_path: impl AsRef<Path>,
+        mut ctx: FunctionEnvMut<'_, WasiEnv>,
+    ) -> Result<ModuleHandle, LinkError> {
+        {
+            let guard = self.state.lock().unwrap();
+
+            if let Some(handle) = guard.side_module_names.get(module_path.as_ref()) {
+                debug_assert!(
+                    guard.side_modules.contains_key(handle),
+                    "Internal error: side module names out of sync with side modules"
+                );
+                return Ok(*handle);
+            }
+        }
+
+        let func_env = ctx.as_ref().clone();
+
+        let (env, mut store) = ctx.data_and_store_mut();
+        let module_bytes = locate_module(module_path.as_ref(), &env.state.fs).await?;
+
+        let module = Module::new(store.engine(), &*module_bytes)?;
+
+        let dylink_info = parse_dylink0_section(&module)?;
+
+        let mut guard = self.state.lock().unwrap();
+        let memory = guard.memory.clone();
+
+        let memory_base = guard.allocate_memory(&memory, &mut store, &dylink_info.mem_info)?;
+        // TODO: handle table allocation... yes, we're even side-stepping that!
+        let table_base = 0;
+
+        let (imports, init) = guard.resolve_imports(
+            &mut store,
+            &func_env,
+            &module,
+            memory.clone(),
+            memory_base,
+            table_base,
+        )?;
+
+        let instance = Instance::new(&mut store, &module, &imports)?;
+
+        let instance_handles =
+            WasiModuleInstanceHandles::new(memory.clone(), &store, instance.clone());
+
+        let loaded_module = DlModule {
+            instance: instance.clone(),
+            memory_base,
+            table_base,
+            instance_handles,
+        };
+
+        let handle = ModuleHandle(guard.next_module_handle);
+        guard.next_module_handle += 1;
```

Comment:
Should we compare the stack offsets here and fail if the module was compiled differently than the main?

Reply:
Dl modules don't have stack offsets, not even the main. We can just configure the stack at will for the main, and sides will use __stack_pointer.